### PR TITLE
Update MonoMod to 25.0.0, add Harmony 2.3 features

### PR DESCRIPTION
**Pull request by Windows10CE**

### CODE REVIEW

File: Harmony/Public/Patching/NativeDetourMethodPatcher.cs [hunk 1]
```diff
@@ -1,176 +0,0 @@
...
-using MonoMod.Utils;
-
-namespace HarmonyLib.Public.Patching
-{
-	/// <summary>
-	///    A method patcher that uses <see cref="MonoMod.RuntimeDetour.NativeDetour" /> to patch internal calls,
-	///    methods marked with <see cref="DynDllImportAttribute" /> and any other managed method that CLR managed-to-native
-	///    trampolines for and which has no IL body defined.
-	/// </summary>
-	public class NativeDetourMethodPatcher : MethodPatcher
```
Comment (ManlyMarco) [inline Harmony/Public/Patching/NativeDetourMethodPatcher.cs:16]: This is a breaking change, though I don't think anyone used this API.

**Review by ManlyMarco — COMMENTED** [review]
I'm not too knowledgeable about the monomod/IL parts that were changed so I'll just leave some comments.

This build doesn't work on BepInEx 5.4.21, preloader fails with
```
System.TypeLoadException: Could not resolve type with token 0100001b (from typeref, class/assembly MonoMod.Utils.Platform, MonoMod.Utils, Version=22.1.29.1, Culture=neutral, PublicKeyToken=null)
  at BepInEx.Preloader.PreloaderRunner.PreloaderPreMain () [0x00000] in <fc9d7fbc6dcb44cf87be11d8d92ae161>:0
...
```

File: Harmony/Public/Patching/PatchManager.cs [hunk 2]
```diff
@@ -143,7 +143,7 @@ internal static MethodBase FindReplacement(StackFrame frame)
 
 			lock (ReplacementToOriginals)
 				return ReplacementToOriginals
-					.FirstOrDefault(kv => kv.Key.IsAlive && ((MethodBase)kv.Key.Target).GetNativeStart().ToInt64() == methodStart).Key.Target as MethodBase;
+					.FirstOrDefault(kv => kv.Key.IsAlive && PlatformTriple.Current.Runtime.GetMethodEntryPoint((MethodBase)kv.Key.Target).ToInt64() == methodStart).Key.Target as MethodBase;
```
Comment (nike4613) [inline Harmony/Public/Patching/PatchManager.cs:146]: This should almost certainly be using the PlatformTriple method rather than the Runtime method.

**Comment by ManlyMarco** [comment]
Also, cd0a86b16f347c9fc13dc96345efb50d3bb234b1 is already merged into master as 4e71dc4f483a8432b001b34e0d2ba7faf47d59b1 so it's not necessary.

File: Harmony/Public/Patching/ManagedMethodPatcher.cs [hunk 3]
```diff
@@ -33,18 +33,17 @@ public override DynamicMethodDefinition PrepareOriginal()
 		/// <inheritdoc />
 		public override MethodBase DetourTo(MethodBase replacement)
 		{
-			ilHook ??= new ILHook(Original, Manipulator, new ILHookConfig {ManualApply = true});
-			// Reset IsApplied to force MonoMod to reapply the ILHook without removing it
-			ILHookExtensions.SetIsApplied(ilHook, false);
```
[the remark this answers is not in the document]
Comment (Windows10CE) [reply Harmony/Public/Patching/ManagedMethodPatcher.cs:38]: ILHookExtensions was using reflection to make MonoMod *think* something wasn't applied when it was, but I'm not sure that would even work under reorg, so I just got rid of the jank and did Undo + Apply instead (which is what this really should've been doing in the first place)

It's a bit slower, but more "correct", and its not like patching was fast anyway

**Review by ManlyMarco — APPROVED** [review]

**Comment by ManlyMarco** [comment]
@Windows10CE I didn't merge this PR yet because it turns out it breaks current BepInEx releases.

**Comment by ManlyMarco** [comment]
I attempted to make BepInEx5 load with this version of HarmonyX but it just hard crashes Unity after I managed to get rid of all the compilation errors https://github.com/BepInEx/BepInEx/commit/ecb4b572e3ae8ec28872d6e2f7a420f91e4b87d9
There's also some APIs straight up missing now, like DynDllImportAttribute.

If you (or anyone else) could made a PR for BepInEx to make it work with this PR without regression then I would be able to merge both.

**Comment by nike4613** [comment]
> This build doesn't work on BepInEx 5.4.21, preloader fails with
> 
> ```
> System.TypeLoadException: Could not resolve type with token 0100001b (from typeref, class/assembly MonoMod.Utils.Platform, MonoMod.Utils, Version=22.1.29.1, Culture=neutral, PublicKeyToken=null)
>   at BepInEx.Preloader.PreloaderRunner.PreloaderPreMain () [0x00000] in <fc9d7fbc6dcb44cf87be11d8d92ae161>:0
> ...
> ```

Looing at this again because somebody else pointed me at it... This looks like the wrong version of Utils. Are you sure you had Reorg's Utils (25.0.0) available to the game?

**Comment by kohanis** [comment]
From what I checked: 
GetExecutingAssembly fix does not work as intended (what's it even for, tho?)
MonoMod.RuntimeDetour 15.x.x really broken on net3.5
MonoMod.RuntimeDetour 15.x.x Hook does not work on mono

Tests on net3.5 only pass because they forward to latest version 4.5+ fm

**Comment by nike4613** [comment]
25 *should* work on .NET 3.5, unless I broke it recently. Mono I'm less sure of, if you can get a testable repro, I can probably fix it pretty quickly.

**Comment by kohanis** [comment]
@nike4613 https://github.com/kohanis/MonoModTest

**Comment by nike4613** [comment]
The `EvenMoreFakeRandom` issue is simply inlining related (.NET 6 and newer don't have an issue in this test because the calling method is not tiered up, so is not compiled with opts), but you're right, .NET 3.5 got broken somewhere along the line. Opened a tracking issue https://github.com/MonoMod/MonoMod/issues/156

**Comment by ManlyMarco** [comment]
I'll wait and see what happens then. if the fix is going to take a long time I can revert this PR.

**Comment by ManlyMarco** [comment]
@nike4613 Any updates? If it looks like the fix is going to take time I'll revert this PR so that others can get merged.

**Comment by nike4613** [comment]
I haven't been working on it much. I'll see what I can figure out today though.

**Comment by Meivyn** [comment]
If someone is willing to try this: https://github.com/Meivyn/HarmonyX/commit/5dc2a46370ea544d6301e45fe1737e739d09eae0

I am unsure what this `Assembly.GetExecutingAssembly` fix is meant for. But since `NativeHook` implementation doesn't seem to support that, I replaced it with a native detour that should fix the Unity crash resulting from its use.

**Comment by nike4613** [comment]
.NET 3.5 is [fixed](https://github.com/MonoMod/MonoMod/commit/b4b4f20fdc00e388609d17d376c99b5f53574219) in the `reorganize` branch now. I've been talking to @Meivyn, and I believe that the reported Mono issues were in fact issues with the implementation in HarmonyX (using `NativeHook` for detouring a managed method).

**Comment by kohanis** [comment]
Yes, in my local repo I used detour to fix it. 
But I also don't understand what fix is even for, because it basically does nothing

**Comment by nike4613** [comment]
Uploaded RuntimeDetour 25.1.0-prerelease.2 with the above fix in it.

**Comment by ManlyMarco** [comment]
@kohanis @Meivyn Can you check if the PR above fixes 3.5 support for you?

**Comment by Meivyn** [comment]
It fixed .NET 3.5 support from what I could see with the repro case, but it still fails with Mono. I found out that the types used in the repro case seem to be related to the issue, as it does work when I replace them with strings.

I do not have a game that makes use of net35 though, so I can't tell how that will affect Harmony.

**Comment by kohanis** [comment]
Original test case works fine on my side with both mono(+unity mono from 2022.3.9f1) and .net framework 3.5, `MonoTest` one was originally wrong usage.
@Meivyn Which mono version fails EvenMoreFakeRandom hook?

**Comment by Meivyn** [comment]
I used the one in Unity 2019.4.28f1.

**Comment by kohanis** [comment]
https://github.com/MonoMod/MonoMod/issues/160
https://github.com/MonoMod/MonoMod/issues/161

**Comment by ManlyMarco** [comment]
Thanks, I'll wait for those to be resolved then.

> (I love NotImplementedException in production btw)

I think it sums Unity up pretty well :)

**Comment by Meivyn** [comment]
Reverse patches seem to be broken somehow. This patch is failing on Harmony 2.11.0, and works fine on 2.10.2: https://github.com/Aeroluna/Heck/blob/master/Heck/HarmonyPatches/PlayViewInterrupter.cs#L76

It throws the exception.

**Comment by ManlyMarco** [comment]
> Reverse patches seem to be broken somehow. This patch is failing on Harmony 2.11.0, and works fine on 2.10.2: https://github.com/Aeroluna/Heck/blob/master/Heck/HarmonyPatches/PlayViewInterrupter.cs#L76
> 
> It throws the exception.

Details? Maybe it adds IL now?

**Comment by Meivyn** [comment]
Problem is that I don't have more details. There's no error nor anything in Harmony logs that could help me figure it out. I'm not familiar with reverse patches, but that `NotImplementedException` at that line is now thrown, and it was not before the update.

**Comment by Meivyn** [comment]
Why this was released without a prior fix of `Assembly.GetExecutingAssembly`?

**Comment by ManlyMarco** [comment]
Reverse patchers appear to work fine so we'd need an exact setup to reproduce the issue.

**Comment by Meivyn** [comment]
Actually the issue might've been caused by my attempt at fixing the `Assembly.GetExecutingAssembly` hook? BSIPA isn't able to launch without a fix for that, so it's hard to tell as I can't try without it.

**Comment by kohanis** [comment]
It looks like inlining. Try calling through a delegate to check, instead of directly

**Pull request merged**